handlers/assign_reviewer/__init__.py

```diff
@@ -20,44 +23,56 @@ def get_collaborators(api):
-    if not match:
-        return None
-    return match.group(1)
-
-# Return a collaborator's username if there is one in the config file. Otherwise, return None.
-def choose_reviewer(api, pull_number):
-    collaborators = get_collaborators(api)
-    if not collaborators:
-        return None
-
-    return collaborators[pull_number % len(collaborators)]
-
-welcome_msg = "Thanks for the pull request, and welcome! The Servo team is excited to review "\
-              "your changes, and you should hear from @%s (or someone else) soon."
+def find_reviewer(comment, collaborators):
+    reviewer = re.search(r'.*r\?[:\- ]*@([a-zA-Z0-9\-]*)', comment)
+    if reviewer and reviewer.group(1) in collaborators:
+        return reviewer.group(1)
+    return None
+
+def get_approver(payload, collaborators):
+    user = payload['comment']['user']['login']
+    comment = payload['comment']['body']
+    approval = re.search(r'.*@bors-servo[: ]*r([\+=])([a-zA-Z0-9\-]*)', comment)
+
+    if approval and user in collaborators:
+        if approval.group(1) == '=':        # check if it's something like "r=username"
+            reviewer = approval.group(2)
+            if reviewer in collaborators:
+                return reviewer
+        return user         # fall back and assign the approver
+    return None
 
 class AssignReviewerHandler(EventHandler):
     def on_pr_opened(self, api, payload):
+        collaborators = get_collaborators(api)
         # If the pull request already has an assignee, don't try to set one ourselves.
         if payload["pull_request"]["assignee"] is not None:
             return
-        reviewer = find_reviewer(payload["pull_request"]["body"]) \
-            or choose_reviewer(api, payload["pull_request"]["number"])
+
+        # find a reviewer from PR comment (like "r? username") or assign one ourselves
+        if collaborators:
+            reviewer = find_reviewer(payload["pull_request"]["body"], collaborators) or \
+                       collaborators[payload["pull_request"]["number"] % len(collaborators)]
+            api.set_assignee(reviewer)
+
         # Add welcome message for new contributors
         author = payload["pull_request"]['user']['login']
         if api.is_new_contributor(author):
-            api.post_comment(welcome_msg % reviewer)
-        api.set_assignee(reviewer)
+            api.post_comment(WELCOME_MSG % reviewer)
 
     def on_new_comment(self, api, payload):
+        collaborators = get_collaborators(api)
         if not self.is_open_pr(payload):
             return
 
-        reviewer = find_reviewer(payload["comment"]["body"])
+        # If this is an approval comment from a reviewer, then assign them
+        reviewer = get_approver(payload, collaborators)
         if reviewer:
             api.set_assignee(reviewer)
+            return
+
+        reviewer = find_reviewer(payload["comment"]["body"], collaborators)
+        api.set_assignee(reviewer)
```

Comment:
This needs a null check: https://github.com/servo/servo/pull/10900#issuecomment-215496739